plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,50 +243,179 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+            }
+        }
+
+        if (command.getPort() != null &&
+                command.getPort() != Integer.parseInt(nuageVspHost.getDetails().get("port"))) {
+            paramsTobeUpdated.put("port", String.valueOf(command.getPort()));
+        }
+
+        GetClientDefaultsCommand getClientDefaultsCmd = new GetClientDefaultsCommand();
+        GetClientDefaultsAnswer getClientDefaultsAnswer = (GetClientDefaultsAnswer) _agentMgr.easySend(nuageVspHost.getId(), getClientDefaultsCmd);
+        String apiVersion = MoreObjects.firstNonNull(command.getApiVersion(), getClientDefaultsAnswer.getCurrentApiVersion());
+        SupportedApiVersionCommand supportedApiVersionCmd = new SupportedApiVersionCommand(apiVersion);
+        SupportedApiVersionAnswer supportedApiVersionAnswer = (SupportedApiVersionAnswer) _agentMgr.easySend(nuageVspHost.getId(), supportedApiVersionCmd);
+        if (!supportedApiVersionAnswer.getResult()) {
+            throw new CloudRuntimeException("Incorrect API version: Nuage plugin only supports " + getClientDefaultsAnswer.getCurrentApiVersion());
+        }
+
+        String apiRelativePath = "/nuage/api/" + command.getApiVersion();
+        if (!apiRelativePath.equals(nuageVspHost.getDetails().get("apirelativepath"))) {
+            paramsTobeUpdated.put("apirelativepath", apiRelativePath);
+            paramsTobeUpdated.put("apiversion", apiVersion);
+        }
 
-                        DetailVO detail = new DetailVO(host.getId(), "nuagevspdeviceid", String.valueOf(nuageVspDevice.getId()));
-                        _hostDetailsDao.persist(detail);
+        if (command.getApiRetryCount() != null &&
+                command.getApiRetryCount() != Integer.parseInt(nuageVspHost.getDetails().get("retrycount"))) {
+            paramsTobeUpdated.put("retrycount", String.valueOf(command.getApiRetryCount()));
+        }
+
+        if (command.getApiRetryInterval() != null &&
+                command.getApiRetryInterval() != Integer.parseInt(nuageVspHost.getDetails().get("retryinterval"))) {
+            paramsTobeUpdated.put("retryinterval", String.valueOf(command.getApiRetryInterval()));
+        }
 
-                        return nuageVspDevice;
+        if (!paramsTobeUpdated.isEmpty()) {
+            Map<String, String> latestParamsValue = new HashMap<String, String>();
+            latestParamsValue.putAll(nuageVspHost.getDetails());
+            latestParamsValue.putAll(paramsTobeUpdated);
+            Map<String, Object> hostdetails = new HashMap<String, Object>(latestParamsValue);
+
+            try {
+                resource.configure("", hostdetails);
+                UpdateNuageVspDeviceCommand cmd = new UpdateNuageVspDeviceCommand(latestParamsValue);
+                UpdateNuageVspDeviceAnswer answer = (UpdateNuageVspDeviceAnswer)_agentMgr.easySend(nuageVspHost.getId(), cmd);
+                if (answer == null || !answer.getResult()) {
+                    s_logger.error("UpdateNuageVspDeviceCommand failed");
```

Comment:
Add context information regarding regarding the operation parameters and device being configured to assist operational debugging.